### ISSUE TITLE
GPS-Pixel coordinate system conversions

### ISSUE THREAD

**Issue by LucasYoung** [NGCP/VTOL#99]
Research NED coordinates.
Convert pixels in image and altitude to ENU or END and then GPS coordinates (given a resolution).
<=>
https://en.wikipedia.org/wiki/Local_tangent_plane_coordinates

**Comment by nmcobb** [comment]
Just realized you'll also need the FOV angle theta.

**Comment by Mr-Recalculate** [comment]
I think i'll need it for both the x and y axis because I assume the camera's sensor will be rectangular.

**Comment by nmcobb** [comment]
Yeah, I think you can calculate the y angle from the aspect ratio?

**Comment by Mr-Recalculate** [comment]
Ok so i'm done for today, but here's a stack overflow page that I used code from and some admittedly terrible implementation that I wrote.
https://stackoverflow.com/questions/7222382/get-lat-long-given-current-point-distance-and-bearing

`from numpy import radians, sin, cos, tan, arctan, hypot, degrees, arctan2, sqrt, pi
import numpy as np

class Meter_to_GPS:
    """
    convert camera distance from orgin (with known GPS coords at orgin) 
    to GPS coords
    """
    
    def __init(self):
        self.R = 6378.1 #km

    def get_bearing_degrees(x, y):
        bear = 0
        if x > 0:
            bear = 90 - arctan(y/x)
        elif x < 0:
            bear = 270 - arctan(y/x)
        elif x == 0:
            if y > 0:
                bear = 0
            elif y < 0:
                bear = 180
            elif y == 0:
                bear = 0 #you haven't moved lol
        return bear

    def get_GPS_coords(self, x, y, Olat, Olon):
        bear = radians(self.bear(x, y))
        d = sqrt((x**2) + (y**2)) #km

        lat = arcsin(sin(Olat)*cos(d/self.R) + cos(Olat)*sin(d/self.R)*cos(bear))
        lon = Olon + arctan2(sin(bear)*sin(d/self.R)*cos(Olat), cos(d/self.R)-sin(Olat)*sin(lat))
        return lat, lon`

**Comment by nmcobb** [comment]
Cool, this could work, some of this is already done for you in conversions.py with the methods gps to ned, and ned to ecef. I can give you more details about these methods on Thursday, but the link I posted above does a decent job of explaining this. these functions should get you the distance in meters.

**Comment by Mr-Recalculate** [comment]
Alright, I've tested this code (pixel coords to meters.) we should be able to use the rest of the methods in the conversions file to get gps coords from pixel coords 

`def pixels2meters(x, y, fov_angles, res, z):
		'''takes x and y as pixel coords, the aspect ration (x and y camera angles) as
		a tuple, the resoution of the camera as a tuple of ints, and z (altitude)'''
		c_x = (tan(np.radians(fov_angles[0])/2)*z*2)/res[0]
		c_y = (tan(np.radians(fov_angles[1])/2)*z*2)/res[1]
		print ("c_x: " + str(c_x))
		print ("c_y: " + str(c_y))
		x = x * c_x
		y = y * c_y
		return x, y`
